learn_init, backwards_flag, example_code

skaae opened the pull request:
Adds the following to Recurrentlayer and LSTM layer:
- Learn init flag
- backwards flag to lstm layer + fixes mask handling when scan is
  going backwards. This means that LSTMlayers with `backwards=True` takes a `mask` input *when get_output_for` is called. Mask is an array with shape=(N_BATCH, LENGTH). 
- adds example with bidirectional LSTM using backwards flag
  -  rename output to hid (LSTM layer), 

I added some ignores to ignore pycharm project files.

We can delete the bidirectional layer and the lstm example. The bidirectional layer does not handle the backwards scan correctly. 

Could we add a new layer without peepholes? Or an option to run without peepholes?

I have a dropout option in my own code. It drops non-recurrent connections. Should we add it? 
I think it is simpler than adding it directly than chaning the code to use layers. I dont see other LSTM implementations having dropout.

Commit:
learn_init, backwards_flag, example_code

Adds the following:
 - Learn init flag to lstmlayer
 - backwards flag to lstm layer + fixes mask handling when scan is
going backwards
 - adds example with bidirectional LSTM using backwards flag

Commit:
cleanup example

Commit:
add backwards+learn_init + example to RNN layer

Adds backwards option to recurrent layer
adds learn_init to recurrent layer
adds bidirectional RNN example

dnouri commented on examples/lstm_backwards.py:
```diff
@@ -0,0 +1,113 @@
+l_out = lasagne.layers.ReshapeLayer(l_recurrent_out,
+                                    (N_BATCH, LENGTH, y_val.shape[-1]))
+
+# Cost function is mean squared error
```
Doesn't look like this comment belongs here.

skaae replied:
nope, I have removed it

dnouri commented on examples/recurrent_bidirectional.py:
```diff
@@ -0,0 +1,144 @@
+# As we iterate over time steps, the input will be batch size x feature dim
+l_recurrent_in_fwd = lasagne.layers.InputLayer(shape=(N_BATCH, X_val.shape[-1]))
+l_input_to_hidden_fwd = lasagne.layers.DenseLayer(l_recurrent_in_fwd, N_HIDDEN,
+                                              nonlinearity=None)
```
This is a minor point about readibility, but I suggest a "from lasagne import layers", which will save you from typing "lasagne" in a lot of places.  Also the arguments are best put in a separate line, your hanging arguments that remove indentation aren't PEP8-compatible.  Just do a line break after opening the bracket and indent by four spaces.  E.g. here:

``` python
l_input_to_hidden_fwd = layers.DenseLayer(
    l_recurrent_in_fwd, N_HIDDEN, nonlinearity=None)
```

skaae replied:
I changed the line breaks, can you look at it?

dnouri replied:
Looks good!  And sorry for being nit-picky; the line breaks are certainly not the most interesting thing going on here.  :-)

skaae replied:
its fine :) I importet layers as well. 
What is your thoughts on:

   1.   inclusion of dropout via a argument: `dropout =0.5`. 
   2.  The bidirectional layer. I think we can remove it?
   3.  I think it has recently been more common to use LSTM's without peepholes. I have an implementation that works. To me the implementation will be messy unless we just add a seperate
LSTM layer without peepholes. Without a seperate layer you need to add if's in init, ,get_parameters, and several places in get_output_for.

dnouri replied:
Just in case you didn't notice: right now this comment is a bit hidden behind this "show outdated diff" thing.  Maybe you want to elevate this to become a proper comment on the pull request.

Your argument regarding point number 3 seems good.  But sadly I've had too little exposure to this code and to LSTMs to be of any real help with the other issues you're asking about.

Commit:
pep8 changes + comments

Commit:
add from lasagne import layers

skaae commented:
its fine :) I importet layers as well. 
What is your thoughts on:
1.   inclusion of dropout via a argument: `dropout =0.5`. Other LSTM's function does not have dropout as far as I know. (http://arxiv.org/pdf/1409.2329.pdf)
2.  The bidirectional layer. I think we can remove it?
3.  I think it has recently been more common to use LSTM's without peepholes. I have an implementation that works. To me the implementation will be messy unless we just add a seperate
   LSTM layer without peepholes. Without a seperate layer you need to add if's in init, ,get_parameters, and several places in get_output_for. ( http://arxiv.org/abs/1409.3215 )

craffel commented:
A couple of things -
- Your example `lstm_backwards.py` does the same thing that `lstm.py` does, just using your code instead.  I've merged it with `lstm.py` and removed it.
- I cleaned up your code for PEP-8 and your docstrings in a few places.
- Can you explain the way your masking works?  Applying the mask directly to the hidden and cell states doesn't make sense to me.

> 1. inclusion of dropout via a argument: dropout =0.5. Other LSTM's function does not have dropout as far as I know. (http://arxiv.org/pdf/1409.2329.pdf)

As discussed here https://github.com/benanne/Lasagne/issues/17#issuecomment-68161719, that will be possible when we create an LSTM layer which can have arbitrary layer connections.  It was pretty unanimous that it was inconsistent/not a good idea to include dropout in the class directly when we have a separate dropout layer.

> 1. The bidirectional layer. I think we can remove it?

Yes, I will remove it once I'm comfortable with the masking scheme you're using.

> 1. I think it has recently been more common to use LSTM's without peepholes. I have an implementation that works. To me the implementation will be messy unless we just add a seperate LSTM layer without peepholes. Without a seperate layer you need to add if's in init, ,get_parameters, and several places in get_output_for. (papers.nips.cc/paper/5426-sequence-to-sequence-learning-with-neural-networks )

I don't like the idea of having a separate "peephole-free" LSTM class because there will be a lot of code duplication, and if we make a change to the way the LSTM layer works, we will have to make the change in two places, which is really bad practice.  I don't think it would be particularly messy to have a peephole connection flag.

skaae commented:
On 03 Jan 2015, at 19:39, Colin Raffel notifications@github.com wrote:

> A couple of things -
> 
> Your example lstm_backwards.py does the same thing that lstm.py does, just using your code instead. I've merged it with lstm.py and removed it.
> Yes. I just wanted to keep the original.
> I cleaned up your code for PEP-8 and your docstrings in a few places.
> Thanks :)
> Can you explain the way your masking works? Applying the mask directly to the hidden and cell states doesn't make sense to me.
> See next post.
> 
> inclusion of dropout via a argument: dropout =0.5. Other LSTM's function does not have dropout as far as I know. (http://arxiv.org/pdf/1409.2329.pdf)
> As discussed here benanne#17 (comment), that will be possible when we create an LSTM layer which can have arbitrary layer connections. It was pretty unanimous that it was inconsistent/not a good idea to include dropout in the class directly when we have a separate dropout layer.
> Ok. It’s just that setting up all these layers is a mess to me.
> The bidirectional layer. I think we can remove it?
> Yes, I will remove it once I'm comfortable with the masking scheme you're using.
> I think it has recently been more common to use LSTM's without peepholes. I have an implementation that works. To me the implementation will be messy unless we just add a seperate LSTM layer without peepholes. Without a seperate layer you need to add if's in init, ,get_parameters, and several places in get_output_for. (papers.nips.cc/paper/5426-sequence-to-sequence-learning-with-neural-networks )
> I don't like the idea of having a separate "peephole-free" LSTM class because there will be a lot of code duplication, and if we make a change to the way the LSTM layer works, we will have to make the change in two places, which is really bad practice. I don't think it would be particularly messy to have a peephole connection flag.
> You need to have seperate step functions, get_parameters, initialization etc. But I can add it.

skaae commented:
The mask is applied with the code:

``` PYTON
            not_mask = 1 - mask
            cell = cell*mask + cell_previous*not_mask
            hid = hid*mask + hid_previous*not_mask
```

not_mask is the negation of mask. When mask is 1 we will return the current hidden state and the current cell state. When mask is 0 we will return the previous hidden state and the previous cell state. When you go backwards the previous hidden and cell states will be inital states until the mask changes to 1. 

I have recently found a similar construction in groundhog: https://github.com/lisa-groundhog/GroundHog/blob/master/groundhog/layers/rec_layers.py#L1483

You dont apply the mask to the hidden state, but ensure that no hidden state is returned before the sequence begins.

Should i Add the no-peephole option?

craffel commented:
> When you go backwards the previous hidden and cell states will be inital states until the mask changes to 1.

Ah I see, so it's essentially a way of carrying the initial state over until the mask flips to 1 (end of sequence is reached when going backwards).  I've added this in as a comment, let me know if it sounds right:
https://github.com/craffel/nntools/commit/8a094f1ddeaef3b3da45d74cc0a8cf13a1d3d9b8#diff-48f9f9d93ed00587a6c49197b4d6e64eR138

> Should i Add the no-peephole option?

Sure, but please do it as a separate pull request.  Make sure you work from the latest commit here though, because there have been a few changes now since your commits.

craffel commented:
Just removed the `BidirectionalLayer`, so all of your changes should be included now.  Thanks for this.

skaae commented:
On 03 Jan 2015, at 20:04, Colin Raffel notifications@github.com wrote:

> When you go backwards the previous hidden and cell states will be inital states until the mask changes to 1.
> Ah I see, so it's essentially a way of carrying the initial state over until the mask flips to 1 (end of sequence is reached when going backwards). I've added this in as a comment, let me know if it sounds right:
> 8a094f1#diff-48f9f9d93ed00587a6c49197b4d6e64eR138
> 
> Exactly :)
> 
> Should i Add the no-peephole option?
> Sure, but please do it as a separate pull request. Make sure you work from the latest commit here though, because there have been a few changes now since your commits.

Ok. Will you merge this commit now? I’ll add the peephole option afterwards. I’m not really sure how to work from your latest commit.
Do you think it’s possible to pass the peephole weights to the step function with  `**kwargs`? or should i create to separate step function.

> — 
> Reply to this email directly or view it on GitHub.

craffel commented:
> Ok. Will you merge this commit now? I’ll add the peephole option afterwards. I’m not really sure how to work from your latest commit.

All commits are merged - the current codebase reflects all your changes and the edits I've made since then:
https://github.com/craffel/nntools/commits/master
You'll need to `git fetch` from this repository and `git merge` to your local master branch.  Or, just create a new fork from the current codebase, whichever is easier for you.

> Do you think it’s possible to pass the peephole weights to the step function with  `**kwargs`? or should i create to separate step function.

Continuing this discussion at https://github.com/benanne/Lasagne/issues/17
